### ISSUE TITLE
Trouble forwarding intercepted packets from TUN interface to default gateway

### ISSUE THREAD

**Issue by Schachte** [songgao/water#101]
I'm on Mac, which I'm realizing is much harder to configure packet forwarding than Linux. 

I created a TUN device and brought it up and assigned it an IP. I also configured the TUN as the default route on my machine. I can successfully intercept packets as expected. However, I want to be able to forward the original packet off to my router so I can still browse the web, etc. 

How might something like this work/look using water? 

I've tried something like 

```go
package main

import (
	"log"
	"net"

	"github.com/songgao/water"
	"golang.org/x/net/ipv4"
)

func main() {
	ifce, err := water.New(water.Config{
		DeviceType: water.TUN,
	})
	if err != nil {
		log.Fatal(err)
	}

	log.Printf("Interface Name: %s\n", ifce.Name())

	packet := make([]byte, 2000)

	// Replace "en0" with your internet-facing network interface name
	internetInterface, err := net.InterfaceByName("en0")
	if err != nil {
		log.Fatal(err)
	}

	conn, err := net.ListenPacket("ip4:4", "0.0.0.0") // Listen for IPv4 packets
	if err != nil {
		log.Fatal(err)
	}
	defer conn.Close()

	rawConn, err := ipv4.NewRawConn(conn)
	if err != nil {
		log.Fatal(err)
	}

	for {
		n, err := ifce.Read(packet)
		if err != nil {
			log.Fatal(err)
		}

		header, err := ipv4.ParseHeader(packet[:n])
		if err != nil {
			log.Printf("Error parsing IPv4 header: %v", err)
			continue
		}

		cm := &ipv4.ControlMessage{
			IfIndex: internetInterface.Index,
		}

		err = rawConn.WriteTo(header, packet[header.Len:n], cm)
		if err != nil {
			log.Printf("Error writing packet to raw socket: %v", err)
		}
	}
}
```

**Comment by songgao** [comment]
I’m not sure simply forwarding the IP packet to a different interface will work because the source IP address would be the one on your TUN. It may also complicate handling any incoming packets. Depending on what you’re trying to do it might be possible to have your user space program act as a NAT router between the TUN consumer and your local gateway, and modify the IP header on the fly accordingly.

**Comment by Schachte** [comment]
@songgao 
Something like this? Trying to make sure I'm thinking of the flow correctly. 
<img width="799" alt="image" src="https://github.com/songgao/water/assets/7055226/231aa415-dcf2-4873-95c4-7e961600e16a">

**Comment by Schachte** [comment]
Ok, I got this working finally.

I created 2 TUN network interfaces on my machine to emulate what you'd typically do in standard VPN implementation, but I'm only testing on a single machine and only using ICMP as an example.

I have my userspace program read off `tun0` IP as expected. I then swap the src/destination IP headers for ICMP requests and recompute the checksum before persisting back to tun interface

```go
func checkSum(data []byte) uint16 {
	var (
		sum    uint32
		length int = len(data)
		index  int
	)
	for length > 1 {
		sum += uint32(data[index])<<8 | uint32(data[index+1])
		index += 2
		length -= 2
	}
	if length > 0 {
		sum += uint32(data[index])
	}
	sum = (sum >> 16) + (sum & 0xffff)
	sum += (sum >> 16)
	return uint16(^sum)
}
```

Persisting back to `tun` in my program has a dest. IP of the other end of the point to point tun and my ping application properly receives the response packet.

I can also simulate a 5 second delay as requested in another Github issue (in my case I simulate a 5 second delay)
<img width="406" alt="image" src="https://github.com/songgao/water/assets/7055226/37247324-9729-496e-a576-c4ac076a364f">
